File: lessons/compiled_lang/go/src/i20.subroutine.go
package main

import "fmt"

var pond = 500 // never mutated - fish() only touches its own local copy
var captured = 0

func fish() {
	pond := 500 // shadows the package-level pond for the rest of this function
	pond -= 150
	captured += 150
}

func main() {
	fmt.Printf("We have %d in this pond.\n", pond)

	fish()
	fmt.Printf("Fishing from a local pond... We now have %d in the main pond.\n", pond)

	fish()
	fmt.Printf("Fishing from a local pond... We now have %d in the main pond.\n", pond)

	fish()
	fmt.Printf("Fishing from a local pond... We now have %d in the main pond.\n", pond)

	fmt.Printf("We now have a total of %d fish captured\n", captured)
}

File: lessons/compiled_lang/cpp/src/i20.subroutine.cpp
#include <iostream>

int pond = 500; // never mutated - fish() only touches its own local copy
int captured = 0;

void fish() {
    int pond = 500; // shadows the global pond for the rest of this function
    pond -= 150;
    captured += 150;
}

int main() {
    std::cout << "We have " << pond << " in this pond.\n";

    fish();
    std::cout << "Fishing from a local pond... We now have " << pond << " in the main pond.\n";

    fish();
    std::cout << "Fishing from a local pond... We now have " << pond << " in the main pond.\n";

    fish();
    std::cout << "Fishing from a local pond... We now have " << pond << " in the main pond.\n";

    std::cout << "We now have a total of " << captured << " fish captured\n";

    return 0;
}
